# Data Cleaning 4: Untagged PER Entities

Strategy 1: Exact Matches

[ Strategy 2: Fuzzy Matches -finally not applied]

In [3]:
from IPython.display import clear_output
import spacy
from spacy import displacy

import pickle

In [2]:
import pandas as pd

In [4]:
file = open("Files_Cleaning3/NER_data_Cleaned3.p", 'rb')
docs = pickle.load(file)

# Exact Matching to Confirmed Names

Let's generate a list of PER entity strings:

In [143]:
peoplestrings =[]
for doc, context in docs:
    for ent in doc.ents:
        if ent.label_=='PER':
            peoplestrings.append(ent.text)         

In [144]:
len(peoplestrings)

32339

In [145]:
peoplestrings = [string.lower() for string in peoplestrings]

In [146]:
len(set(peoplestrings))

12805

In [147]:
peoplestrings = list(set(peoplestrings))

In [11]:
untaggedtext

' Pago de'

In [12]:
strlist

['fernando', 'de', 'ledesma']

In [13]:
untaggedtokens

['Pago', 'de']

In [14]:
untaggedtokens[start+b]

'de'

In [15]:
b

0

In [16]:
span

''

In [ ]:
entry

In [ ]:
untaggedPERS = []

for doc, context in docs:
    ents = list(doc.ents)
    enttexts = []
    for ent in ents:
        enttexts.append(ent.text)
        untaggedtext = doc.text.replace(ent.text," ",1) #replace the entity span once with a blank
    for string in peoplestrings:
        strlist = string.split()
        try:
            untaggedtokens= untaggedtext.split()
            for i in range(len(untaggedtokens)):
                start = i
                b=0
                while (b<len(strlist)) and (untaggedtokens[start+b].lower() == strlist[b].lower()):
                    b= b+1
        except:
            pass
        end = start + b
        span = ' '.join(untaggedtokens[start:end])
        if (b !=0) and (span.lower() == string.lower()):
            docid = context['id']
            entry = [docid, string, span]
            untaggedPERS.append(entry)
            b=0
    print(context['id'])
    clear_output(wait=True)
        
#        if (string.lower() in doc.text.lower()) and (string.lower() not in enttexts):
 #           string = str(string)
  #          docid = context['id']
   #         entry = [docid, string]
    #        untaggedPERS.append(entry)

In [148]:
untaggedPERS = []

for doc, context in docs:
    untaggedtext = ''
    for token in doc:
        if token.ent_iob==2:
            untaggedtext = untaggedtext + " " + token.text
    for string in peoplestrings:
        b=0
        strlist = string.split()
        try:
            untaggedtokens= untaggedtext.split()
            for i in range(len(untaggedtokens)):
                start = i
                b=0
                while (b<=len(strlist)) and (untaggedtokens[start+b].lower() == strlist[b].lower()):
                    b= b+1
        except:
            pass
        end = start + b
        span = ' '.join(untaggedtokens[start:end])
        if (b !=0) and (len(range(start,end)) == len(strlist)):
            docid = context['id']
            entry = [docid, string, span, start, end]
            untaggedPERS.append(entry)
    print(context['id'])
    clear_output(wait=True)

8607


In [40]:
untaggedtokens[start:end]

[]

In [29]:
range(len(untaggedtokens))

range(0, 2)

In [27]:
span

''

In [17]:
strlist

['fernando', 'de', 'ledesma']

In [18]:
len(strlist)

3

In [ ]:
len(docs)

In [38]:
docs[99][1]

{'id': 100}

In [39]:
displacy.render(docs[99][0], style ='ent')

In [ ]:
#Need to remove duplicates with capitalization variations. Need to make sure spans are not double counted.

In [149]:
len(untaggedPERS)

2476

In [150]:
untaggedPERS_df = pd.DataFrame(untaggedPERS, columns = ['docid','match','string','start','end'])

In [151]:
untaggedPERS_df

docid             match            string  start   end
0         2             costa             costa     62    63
1        52             costa             costa   1030  1031
2        55             costa             costa   1004  1005
3        82             costa             costa    269   270
4        85              juan              juan     11    12
...     ...               ...               ...    ...   ...
2471   8513        juan de la        Juan de la      3     6
2472   8513  juan de la palma  Juan de la Palma      3     7
2473   8536            corral            corral     36    37
2474   8552            miguel            Miguel    687   688
2475   8562             rubio             rubio     33    34

[2476 rows x 5 columns]

In [152]:
untaggedPERS_df.to_csv('Files_Cleaning4/untaggedPERstrings_strat1.csv')

In [153]:
untaggedPERS_df[untaggedPERS_df['string']!='de']

docid             match            string  start   end
0         2             costa             costa     62    63
1        52             costa             costa   1030  1031
2        55             costa             costa   1004  1005
3        82             costa             costa    269   270
4        85              juan              juan     11    12
...     ...               ...               ...    ...   ...
2471   8513        juan de la        Juan de la      3     6
2472   8513  juan de la palma  Juan de la Palma      3     7
2473   8536            corral            corral     36    37
2474   8552            miguel            Miguel    687   688
2475   8562             rubio             rubio     33    34

[2476 rows x 5 columns]

In [154]:
untaggedPERstrings = pd.DataFrame(untaggedPERS_df['string'].value_counts())

In [155]:
untaggedPERstrings.to_csv('Files_Cleaning4/untaggedPERstrings_strat1_unique.csv')

In [156]:
untaggedPERstrings[untaggedPERstrings.index.str.contains(' ')]

string
juan bautista            35
señor prouisor           24
juan baptista            17
Juan Bautista            15
francisco de             13
...                     ...
geronimo belasques        1
justa rufina              1
Geronima de çamudio       1
Diego de Alcala           1
hernan Ruys               1

[147 rows x 1 columns]

Looking at the contents of this DataFrame in the .csv file, we have our work cut out for us. But we can remove matches with the strings "de", "costa", and "de la".

In [157]:
untaggedPERS_df = untaggedPERS_df[(untaggedPERS_df['string']!='de') & (untaggedPERS_df['string']!='de la') & (untaggedPERS_df['string']!='costa')]

In [158]:
len(untaggedPERS_df)

2042

In [160]:
untaggedPERS_df['addentity']=''
untaggedPERS_df['entstart']=''
untaggedPERS_df['entend']=''
untaggedPERS_df['label']=''

/Users/Felipe/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/Users/Felipe/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/Felipe/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_index

In [161]:
untaggedPERS_df.head()

docid           match          string  start  end addentity label entstart  \
4     85            juan            juan     11   12                            
5     85          juan\n            juan     11   12                            
6     85         juan de         juan de     11   13                            
7     85  señor prouisor  señor prouisor     23   25                            
8     85      juan de la      juan de la     11   14                            

  entend  
4         
5         
6         
7         
8

Some potential entities have matched multiple entity strings. We need to take the longest spans and keep those. To do this, we sort on docid, start, and end, in descending order, so the longest spans are first (ascending = False). Then, we use  df.drop_duplicates with keep =  first.

In [176]:
untaggedPERS_df = untaggedPERS_df.sort_values(['docid','start','end'], ascending = False).drop_duplicates(['docid','start'], keep = 'first')

In [177]:
untaggedPERS_df

docid             match            string  start  end addentity label  \
2475   8562             rubio             rubio     33   34                   
2474   8552            miguel            Miguel    687  688                   
2473   8536            corral            corral     36   37                   
2472   8513  juan de la palma  Juan de la Palma      3    7                   
2466   8507          bautista          Bautista     73   74                   
...     ...               ...               ...    ...  ...       ...   ...   
15      100             maria             maria     82   83                   
12       95               ana               ana     55   56                   
11       93             roque             roque     48   49                   
7        85    señor prouisor    señor prouisor     23   25                   
9        85  juan de la palma  juan de la palma     11   15                   

     entstart entend  
2475                  
2474                  
2473                  
2472                  
2466                  
...       ...    ...  
15                    
12                    
11                    
7                     
9                     

[1624 rows x 9 columns]

That leaves 1,624 potentially untagged PER entities to look at. 

In [178]:
untaggedPERS_df.to_csv('Files_Cleaning4/untaggedPERstrings_strat1_clean.csv')

In [1]:
import pandas as pd
untaggedPERS_df = pd.read_csv('Files_Cleaning4/untaggedPERstrings_strat1_manual.csv', index_col=0)

In [2]:
untaggedPERS_df = untaggedPERS_df.fillna('')

In [12]:
count = 0
for (index, row) in untaggedPERS_df.iterrows():
    if  (row['addentity']==''):
        count=count+1
        print(count)
        print(row['string'], row['match'])
        docid = row['docid']
        if ' ' in row['string']:
            strlist = row['string'].split()
        else: 
            strlist = [row['string']]
        for doc, context in docs:
            if context['id'] == docid: 
                for token in doc:
                    b=0
                    if token.ent_iob == 2 and token.text.lower()==strlist[0].lower():
                        start = token.i
                        while (b<len(strlist)) and (str(doc[start+b].text).lower() == strlist[b].lower()):
                            b= b+1
                        end = start + b
                        span = doc[start:end]
                        print(span)
                        if (b !=0) and (len(range(start,end)) == len(strlist)):
                            try:
                                displacy.render(doc[start-5:end+5],style='ent',jupyter=True)
                            except:
                                displacy.render(doc[start:end+5],style='ent',jupyter=True)
                            untaggedPERS_df.loc[index, 'addentity']=input('Add?')
                        clear_output(wait=True)

1
francisco de francisco de
francisco


Note: in this notebook I am only marking PER entities as 1. Some of these entities are clearly LOC and ORG entities, but these will be better found with LOC and ORG strings. 

In [3]:
untaggedPERS_df.groupby('addentity').count()

Unnamed: 0.1  docid  match  string  start   end  label  entstart  \
addentity                                                                     
0                  1263   1263   1263    1263   1263  1263      0         0   
1                   342    342    342     342    342   342      0         0   
len                  18     18     18      18     18    18      0         0   

           entend  
addentity          
0               0  
1               0  
len             0

In [10]:
untaggedPERS_df.to_csv('Files_Cleaning4/untaggedPERstrings_strat1_manual.csv')

In [ ]:
# Now make the edits (go through all value 1, add spans to newstart newend because some need edits)

In [6]:
untaggedPERS_df['newstart']=''
untaggedPERS_df['newend']=''

In [22]:
# untaggedPERS_df[untaggedPERS_df['addentity']=='e']='1'

In [21]:
count = 0
for (index, row) in untaggedPERS_df.iterrows():
    if (row['addentity']==''):
        docid = row['docid']
        if ' ' in row['string']:
            strlist = row['string'].split()
        else: 
            strlist = [row['string']]
        for doc, context in docs:
            if context['id'] == docid: 
                for token in doc:
                    b=0
                    if token.ent_iob == 2 and token.text.lower()==strlist[0].lower():
                        count=count+1
                        print(count)
                        start = token.i
                        while (b<len(strlist)) and (str(doc[start+b].text).lower() == strlist[b].lower()):
                            b= b+1
                        end = start + b
                        span = doc[start:end]
                        print(start, end, span)
                        if len(strlist)==len(range(start,end)):
                            try:
                                displacy.render(doc[start-5:end+5],style='ent',jupyter=True)
                            except:
                                displacy.render(doc[start:end+5],style='ent',jupyter=True)
                            c = input('See more?')
                            if c == 'y':
                                displacy.render(doc, style='ent', jupyter=True)
                            a=input('New start?')
                            b=input('New end?')
                            d= input('Add?')           
                            if d =='e':
                                untaggedPERS_df.loc[index,'addentity']='e'
                                if a !='':
                                    untaggedPERS_df.loc[index, 'newstart']=a
                                if b !='':
                                    untaggedPERS_df.loc[index,'newend']=b
                            else:
                                untaggedPERS_df.loc[index,'addentity']=d
                    
                            clear_output(wait=True)

1
135 136 francisco


In [22]:
untaggedPERS_df.groupby('addentity').count()

Unnamed: 0.1  docid  match  string  start   end  label  entstart  \
addentity                                                                     
                      1      1      1       1      1     1      1         1   
0                  1296   1296   1296    1296   1296  1296   1296      1296   
e                   309    309    309     309    309   309    309       309   
len                  18     18     18      18     18    18     18        18   

           entend  newstart  newend  
addentity                            
                1         1       1  
0            1296      1296    1296  
e             309       309     309  
len            18        18      18

In [18]:
untaggedPERS_df.to_csv('Files_Cleaning4/untaggedPERstrings_strat1_manual.csv')

In [23]:
untagged_PERSedits = untaggedPERS_df[untaggedPERS_df['addentity']=='e']

In [24]:
untagged_PERSedits

Unnamed: 0.1  docid              match             string  start  end  \
40            2421   8451             alonso             Alonso     82   83   
73            2363   7838  conde de arenales  conde de Arenales     33   36   
97            2336   7341            domingo            Domingo     11   12   
99            2332   7232            morillo            MORILLO     12   13   
106           2324   7190            murillo            MURILLO     22   23   
...            ...    ...                ...                ...    ...  ...   
1585            76    294     señor prouisor     señor prouisor    142  144   
1602            43    234            josepha            josepha     21   22   
1616            18    101              roque              roque    261  262   
1617            17    101             ortega             ortega    150  151   
1622             7     85     señor prouisor     señor prouisor     23   25   

     addentity label entstart entend newstart newend  
40           e                            101    102  
73           e                             37     40  
97           e                             20     21  
99           e                             16     17  
106          e                             30     31  
...        ...   ...      ...    ...      ...    ...  
1585         e                            360    362  
1602         e                             52     57  
1616         e                            297    298  
1617         e                            186    187  
1622         e                             25     27  

[309 rows x 12 columns]

First, I will remove any tags that are completely duplicated (based on document, newstart and newend)

In [40]:
untagged_PERSedits = untagged_PERSedits.drop_duplicates(['docid','newstart','newend'])

There are also a few other repeated ones, we should take the longest ones. 

In [51]:
untagged_PERSedits[untagged_PERSedits.duplicated(subset=['docid','newend'], keep=False)].sort_values(['docid','start','end'], ascending = True)

Empty DataFrame
Columns: [Unnamed: 0.1, docid, match, string, start, end, addentity, label, entstart, entend, newstart, newend]
Index: []

In [64]:
untagged_PERSedits_list = untagged_PERSedits[['docid','newstart','newend']].to_numpy().tolist()

Before adding them, I checked the spans visually to ensure none overlap. 

In [67]:
untagged_PERSedits_list

[[85, '25', '27'],
 [101, '186', '187'],
 [101, '297', '298'],
 [234, '52', '57'],
 [294, '360', '362'],
 [300, '111', '113'],
 [335, '159', '162'],
 [423, '7', '8'],
 [590, '21', '22'],
 [678, '483', '484'],
 [798, '28', '32'],
 [870, '94', '96'],
 [893, '40', '43'],
 [900, '190', '191'],
 [900, '194', '196'],
 [903, '294', '296'],
 [916, '6', '9'],
 [923, '30', '32'],
 [932, '537', '540'],
 [949, '660', '662'],
 [953, '269', '270'],
 [969, '116', '118'],
 [975, '192', '194'],
 [1012, '45', '48'],
 [1014, '3276', '3279'],
 [1014, '3411', '3414'],
 [1022, '435', '438'],
 [1027, '2', '5'],
 [1066, '1366', '1368'],
 [1072, '39', '41'],
 [1100, '0', '2'],
 [1105, '28', '29'],
 [1121, '52', '53'],
 [1199, '2262', '2264'],
 [1200, '3179', '3182'],
 [1201, '948', '951'],
 [1212, '1676', '1677'],
 [1212, '2105', '2107'],
 [1218, '497', '499'],
 [1225, '259', '262'],
 [1227, '692', '695'],
 [1229, '947', '950'],
 [1294, '165', '168'],
 [1340, '7', '9'],
 [1355, '2304', '2305'],
 [1356, '159', 

In [84]:
for line in range(len(untagged_PERSedits_list)):
    untagged_PERSedits_list[line] = [int(i) for i in untagged_PERSedits_list[line]]

In [86]:
untagged_PERSedits_list

[[85, 25, 27],
 [101, 186, 187],
 [101, 297, 298],
 [234, 52, 57],
 [294, 360, 362],
 [300, 111, 113],
 [335, 159, 162],
 [423, 7, 8],
 [590, 21, 22],
 [678, 483, 484],
 [798, 28, 32],
 [870, 94, 96],
 [893, 40, 43],
 [900, 190, 191],
 [900, 194, 196],
 [903, 294, 296],
 [916, 6, 9],
 [923, 30, 32],
 [932, 537, 540],
 [949, 660, 662],
 [953, 269, 270],
 [969, 116, 118],
 [975, 192, 194],
 [1012, 45, 48],
 [1014, 3276, 3279],
 [1014, 3411, 3414],
 [1022, 435, 438],
 [1027, 2, 5],
 [1066, 1366, 1368],
 [1072, 39, 41],
 [1100, 0, 2],
 [1105, 28, 29],
 [1121, 52, 53],
 [1199, 2262, 2264],
 [1200, 3179, 3182],
 [1201, 948, 951],
 [1212, 1676, 1677],
 [1212, 2105, 2107],
 [1218, 497, 499],
 [1225, 259, 262],
 [1227, 692, 695],
 [1229, 947, 950],
 [1294, 165, 168],
 [1340, 7, 9],
 [1355, 2304, 2305],
 [1356, 159, 161],
 [1366, 470, 472],
 [1367, 677, 679],
 [1373, 792, 794],
 [1400, 7, 10],
 [1444, 79, 81],
 [1466, 875, 877],
 [1466, 1075, 1078],
 [1486, 227, 231],
 [1492, 1069, 1071],
 [1506

This is the list of untagged entities to be incorporated to our spacy file. 

# Incorporating 

In [87]:
for doc, context in docs:
    for line in untagged_PERSedits_list:
        if line[0] == context['id']:
            a = spacy.tokens.Span(doc, int(line[1]),int(line[2]),label = 'PER')
            doc.ents = list(doc.ents) + [a]
            print(a)

señor prouisor
ortega
roque
theresa josepha calista de jesus
señor prouisor
señor Prouisor
phelipe del castillo
Rodriguez
Juan
isabel
favian de umanes canzino
marco caña
marques de tarifa
Alfian
A. Rodriguez
señor prouisor
marques de tarifa
señor prouisor
— vasco perea
hernan ruyz
francisco
hernan ruys
hernan ruyz
grauiel del valle
Fernando Martin Robles
grabiel del valle
grabiel del balle
baltasar de riuera
placido pacheco
señor prouisor
.«juan franco
ysabel
Legott
rubricado)-Pedro diaz
rubricado)-francisco del carpio
rubricado)-Francisco de Vargas
geronima
rubricado)-matheo diaz
bias martin
rubricado)-pedro diaz despina
felis de obiedo
luys de peña
frutos de villacastin
miguel a
montañes
señor prouisor
señor prouisor
señor prouisor
señor prouisor
phelipe de ribas
señor prouisor
miguell guelles
bernabe de baeza
hartolome de la puerta
rubricado)=alonso cano
señor prouisor
pero fernandez
luysa de beatriz
Francisca
señor prouisor
señor prouisor
josephe de arze
pedro criste
Alonso esijano

To double check, the number of entities should now be 296 + 71193:

In [88]:
count =0
for doc,context in docs:
    ents = list(doc.ents)
    for ent in ents:
        count = count+1

In [89]:
count

71489

In [90]:
count - 71193

296

In [91]:
pickle.dump(docs, open( "Files_Cleaning4/NER_data_Cleaned4.p", "wb" ))